backend/stellar_refund_validator.go

```diff
@@ -0,0 +1,111 @@
+package backend
+
+import (
+	"context"
+	"database/sql"
+	"net/http"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/render/problem"
+
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/starbridge/ethereum"
+	"github.com/stellar/starbridge/store"
+)
+
+var RefundAlreadyExecuted = problem.P{
+	Type:   "refund_already_executed",
+	Title:  "Refund Already Executed",
+	Status: http.StatusBadRequest,
+	Detail: "The refund has already been executed.",
+}
+
+// StellarRefundValidator checks if it is possible to
+// refund a deposit to depositor's Stellar account.
+type StellarRefundValidator struct {
+	Session                db.SessionInterface
+	WithdrawalWindow       time.Duration
+	Observer               ethereum.Observer
+	EthereumFinalityBuffer uint64
+}
+
+// StellarRefundDetails includes metadata about the
+// validation result.
+type StellarRefundDetails struct {
+	// LedgerSequence is the sequence number of the Stellar ledger
+	// for which the validation result is accurate.
+	LedgerSequence uint32
+}
+
+func (s StellarRefundValidator) CanRefund(ctx context.Context, deposit store.StellarDeposit) (StellarRefundDetails, error) {
+	dbStore := store.DB{Session: s.Session.Clone()}
+	err := dbStore.Session.BeginTx(&sql.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  true,
+	})
+	if err != nil {
+		return StellarRefundDetails{}, errors.Wrap(err, "error starting repeatable read transaction")
+	}
+	defer func() {
+		// explicitly ignore return value to make the linter happy
+		_ = dbStore.Session.Rollback()
+	}()
+
+	lastLedgerSequence, err := dbStore.GetLastLedgerSequence(context.Background())
+	if err != nil {
+		return StellarRefundDetails{}, errors.Wrap(err, "error getting last ledger sequence")
+	}
+
+	// Checks on Stellar side:
```

Comment:
it's not necessary to check the withdrawal window condition on the source chain (in this case stellar). We only need to do it on the destination chain